File: 2653.sliding-subarray-beauty.py
#
# @lc app=leetcode id=2653 lang=python3
#
# [2653] Sliding Subarray Beauty
#

# @lc code=start]
from sortedcontainers import SortedList
class Solution:
    def getSubarrayBeauty(self, nums: list[int], k: int, x: int) -> list[int]:
        #! This is the brutal force

        #todo sub-problem:
        #todo 1. In the subarray, find the number of negative integers

        #todo 2. In the subarray, find the order of the numbers. where I need to fix it by iterating the subarray

        #! The key takeaway is to see that the ans will always be either 0 or negative numbers I have to improve on exploring how the problem conditions would affect the output

        # ans = []        
        # #!check whether the element is negative or not
        # freq = [0] * 50
        # for i in range(len(nums)):
        #     #Enter the window
        #     if nums[i] < 0:
        #         freq[50 + nums[i]] += 1   
        #     if i < k - 1:
        #         continue
        #     #! Record the number of negative integers
        #     count = 0
        #     for j in range(50):
        #         count += freq[j]
        #         if count >=x:
        #             ans.append(j - 50)
        #             break
        #     if count < x:
        #         ans.append(0) 
        #     #Exit the window
        #     if nums[i - k + 1] < 0:
        #         freq[50 + nums[i - k + 1]] -= 1
        # return ans

        #! This is the most efficient soluton 91.53% faster than all users, using the sorted list method
        # neg_list = SortedList()
        # ans = []
        # neg_count = 0
        # for i in range(len(nums)):
        #     if nums[i] < 0:
        #         neg_list.add(nums[i])
        #         neg_count += 1
        #     if i < k - 1:
        #         continue
        #     if neg_count >= x:
        #         ans.append(neg_list[x - 1])
        #     else: 
        #         ans.append(0)
        #     #Exit the window
        #     if nums[i - k + 1] < 0:
        #         neg_list.remove(nums[i - k + 1])
        #         neg_count -= 1
        # return ans

        ans = []
        freq = [0] * 50
        for i in range(len(nums)):
            if nums[i] < 0:
                freq[50 + nums[i]] += 1
            if i < k -1:
                continue

            cnt = 0
            for j in range(50):
                cnt += freq[j]
                if cnt >= x:
                    ans.append(j - 50)
                    break
            if cnt < x:
                ans.append(0)
            if nums[i - k + 1] < 0:
                freq[50 + nums[i - k + 1]] -= 1
        return ans
# @lc code=end

File: 2653.sliding-subarray-beauty.java
/*
 * @lc app=leetcode id=2653 lang=java
 *
 * [2653] Sliding Subarray Beauty
 */
import java.util.*;
// @lc code=start
class Solution {
    PriorityQueue<Integer> minpq;
    PriorityQueue<Integer> maxpq;
    int[] location;
    int maxpqHave;
    int minpqHave;
    public int[] getSubarrayBeauty(int[] nums, int k, int x) {
        minpq = new PriorityQueue<>(Comparator.comparingInt(a -> nums[a]));
        maxpq = new PriorityQueue<>((a,b)-> nums[b] - nums[a]);
        location = new int[nums.length];// 0为删除 1在大堆 -1在小堆
        maxpqHave = 0; // 大堆有效元素
        minpqHave = 0; // 小堆有效元素

        int[] ans = new int[nums.length - k + 1];
        for (int i = 0; i < nums.length; i++) {
            // 入窗口
            if (nums[i] < 0){
                maxpq.add(i);// 不管什么元素，只要符合条件(<0)先入大顶堆
                location[i] = 1;
                maxpqHave++;
            }
            if (i >= k && nums[i - k] < 0){
                //延迟删除
                if (location[i - k] < 0){
                    minpqHave--;
                }else {
                    maxpqHave--;
                }
                location[i - k] = 0; // 标记删除
            }
            if (i >= k - 1){
                ans[i - k + 1] = pqGet(x, nums);
            }
        }
        return ans;
    }

    private int pqGet(int x, int[] nums) {
        pqAdjust(nums);
        while (maxpqHave > x){
            pqRemove(maxpq);
            Integer poll = maxpq.poll();
            minpq.add(poll);
            maxpqHave--;
            minpqHave++;
            location[poll] = -1;
        }
        while (maxpqHave < x){
            pqRemove(minpq);
            if (minpqHave > 0){
                Integer poll = minpq.poll();
                maxpq.add(poll);
                maxpqHave++;
                minpqHave--;
                location[poll] = 1;
            }else if (maxpqHave < x){
                break;
            }
        }
        pqRemove(maxpq);
        if (maxpqHave == x){
            return nums[maxpq.peek()];
        }
        return 0;
    }
    private void pqAdjust(int[] nums) {
        while (maxpqHave > 0 && minpqHave > 0 && nums[maxpq.peek()] > nums[minpq.peek()]){
            pqRemove(maxpq);
            pqRemove(minpq);
            Integer poll = maxpq.poll();
            minpq.add(poll);
            location[poll] = -1;
            maxpqHave--;
            minpqHave++;
        }
    }

    private void pqRemove(PriorityQueue<Integer> pq) {
        while (!pq.isEmpty() && location[pq.peek()] == 0){
            pq.poll();
        }
    }
}
// @lc code=end

